{
 "cells": [
  {
   "cell_type": "code",
   "id": "50327de0-144b-4019-a5b0-72ba84148530",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-30T04:15:07.875263Z",
     "start_time": "2024-08-30T04:15:06.167051Z"
    }
   },
   "source": [
    "pip install pandas"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in /home/wojtek/anaconda3/envs/enhancerlit/lib/python3.11/site-packages (2.2.2)\r\n",
      "Requirement already satisfied: numpy>=1.23.2 in /home/wojtek/anaconda3/envs/enhancerlit/lib/python3.11/site-packages (from pandas) (1.26.4)\r\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/wojtek/anaconda3/envs/enhancerlit/lib/python3.11/site-packages (from pandas) (2.9.0.post0)\r\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/wojtek/anaconda3/envs/enhancerlit/lib/python3.11/site-packages (from pandas) (2024.1)\r\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/wojtek/anaconda3/envs/enhancerlit/lib/python3.11/site-packages (from pandas) (2024.1)\r\n",
      "Requirement already satisfied: six>=1.5 in /home/wojtek/anaconda3/envs/enhancerlit/lib/python3.11/site-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\r\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "cell_type": "code",
   "id": "93fdace3-ad1a-4650-8077-a73415c92349",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-30T04:15:12.291674Z",
     "start_time": "2024-08-30T04:15:12.020642Z"
    }
   },
   "source": [
    "import requests\n",
    "from xml.etree import ElementTree\n",
    "import pandas as pd\n",
    "\n",
    "def pubmed_search(query, retmax=10):\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\"\n",
    "    params = {\n",
    "        \"db\": \"pubmed\",\n",
    "        \"term\": query,\n",
    "        \"retmode\": \"xml\",\n",
    "        \"retmax\": retmax\n",
    "    }\n",
    "    \n",
    "    response = requests.get(base_url, params=params)\n",
    "    response.raise_for_status()\n",
    "    \n",
    "    tree = ElementTree.fromstring(response.content)\n",
    "    ids = [id_elem.text for id_elem in tree.findall(\".//Id\")]\n",
    "    \n",
    "    return ids\n",
    "\n",
    "def fetch_details(id_list):\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    ids = \",\".join(id_list)\n",
    "    params = {\n",
    "        \"db\": \"pubmed\",\n",
    "        \"id\": ids,\n",
    "        \"retmode\": \"xml\",\n",
    "        \"rettype\": \"abstract\"\n",
    "    }\n",
    "    \n",
    "    response = requests.get(base_url, params=params)\n",
    "    response.raise_for_status()\n",
    "    \n",
    "    return response.content"
   ],
   "outputs": [],
   "execution_count": 2
  },
  {
   "cell_type": "code",
   "id": "424e46d1-3d77-46a2-9838-8aee4ddfbda1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-30T04:15:13.032962Z",
     "start_time": "2024-08-30T04:15:13.028128Z"
    }
   },
   "source": [
    "def preprocess_details_to_dataframe(xml_data):\n",
    "    tree = ElementTree.fromstring(xml_data)\n",
    "    articles = []\n",
    "\n",
    "    for article in tree.findall(\".//PubmedArticle\"):\n",
    "        article_data = {}\n",
    "        \n",
    "        # Extracting the title\n",
    "        title_elem = article.find(\".//ArticleTitle\")\n",
    "        article_data['Title'] = title_elem.text if title_elem is not None else \"N/A\"\n",
    "        \n",
    "        # Extracting the authors\n",
    "        authors = []\n",
    "        for author in article.findall(\".//Author\"):\n",
    "            last_name = author.find(\"LastName\")\n",
    "            fore_name = author.find(\"ForeName\")\n",
    "            if last_name is not None and fore_name is not None:\n",
    "                authors.append(f\"{fore_name.text} {last_name.text}\")\n",
    "        article_data['Authors'] = \", \".join(authors) if authors else \"N/A\"\n",
    "        \n",
    "        # Extracting the journal name\n",
    "        journal_elem = article.find(\".//Journal/Title\")\n",
    "        article_data['Journal'] = journal_elem.text if journal_elem is not None else \"N/A\"\n",
    "        \n",
    "        # Extracting the publication date\n",
    "        pub_date_elem = article.find(\".//PubDate\")\n",
    "        if pub_date_elem is not None:\n",
    "            year_elem = pub_date_elem.find(\"Year\")\n",
    "            month_elem = pub_date_elem.find(\"Month\")\n",
    "            day_elem = pub_date_elem.find(\"Day\")\n",
    "            pub_date = f\"{year_elem.text if year_elem is not None else ''}-{month_elem.text if month_elem is not None else ''}-{day_elem.text if day_elem is not None else ''}\"\n",
    "            article_data['Publication Date'] = pub_date\n",
    "        else:\n",
    "            article_data['Publication Date'] = \"N/A\"\n",
    "        \n",
    "        # Extracting the abstract\n",
    "        abstract_elem = article.find(\".//Abstract/AbstractText\")\n",
    "        article_data['Abstract'] = abstract_elem.text if abstract_elem is not None else \"N/A\"\n",
    "        \n",
    "        articles.append(article_data)\n",
    "\n",
    "    df = pd.DataFrame(articles)\n",
    "    return df"
   ],
   "outputs": [],
   "execution_count": 3
  },
  {
   "cell_type": "code",
   "id": "ca323139-6c36-438b-b6fa-7fe74e8248d9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-30T04:15:13.813107Z",
     "start_time": "2024-08-30T04:15:13.810683Z"
    }
   },
   "source": [
    "def fetch_full_text(pmc_id):\n",
    "    base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "    params = {\n",
    "        \"db\": \"pmc\",\n",
    "        \"id\": pmc_id,\n",
    "        \"retmode\": \"xml\"\n",
    "    }\n",
    "    \n",
    "    response = requests.get(base_url, params=params)\n",
    "    response.raise_for_status()\n",
    "    \n",
    "    return response.content"
   ],
   "outputs": [],
   "execution_count": 4
  },
  {
   "cell_type": "code",
   "id": "63201f21-301c-4c78-b878-f0d21c2862ac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-30T04:15:14.934437Z",
     "start_time": "2024-08-30T04:15:14.931251Z"
    }
   },
   "source": [
    "def preprocess_full_text_to_plain_text(xml_data):\n",
    "    tree = ElementTree.fromstring(xml_data)\n",
    "    text_content = []\n",
    "\n",
    "    # Extract text from <body> element of the article\n",
    "    body_elem = tree.find(\".//body\")\n",
    "    if body_elem is not None:\n",
    "        for elem in body_elem.iter():\n",
    "            if elem.text:\n",
    "                text_content.append(elem.text.strip())\n",
    "            if elem.tail:\n",
    "                text_content.append(elem.tail.strip())\n",
    "    \n",
    "    plain_text = \"\\n\".join(text_content)\n",
    "    return plain_text"
   ],
   "outputs": [],
   "execution_count": 5
  },
  {
   "cell_type": "code",
   "id": "d16363c4-02c6-4cac-8a78-7093d2a3ec72",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-30T04:15:16.422925Z",
     "start_time": "2024-08-30T04:15:16.419177Z"
    }
   },
   "source": [
    "def fetch_text_from_pubmed_id(id):\n",
    "    # Convert PubMed ID to PMC ID using eLink\n",
    "    elink_base_url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi\"\n",
    "    elink_params = {\n",
    "        \"dbfrom\": \"pubmed\",\n",
    "        \"db\": \"pmc\",\n",
    "        \"id\": id,\n",
    "        \"retmode\": \"xml\"\n",
    "    }\n",
    "    \n",
    "    elink_response = requests.get(elink_base_url, params=elink_params)\n",
    "    elink_response.raise_for_status()\n",
    "    \n",
    "    elink_tree = ElementTree.fromstring(elink_response.content)\n",
    "    pmc_id_elem = elink_tree.find(\".//LinkSetDb/Link/Id\")\n",
    "    if pmc_id_elem is not None:\n",
    "        pmc_id = pmc_id_elem.text\n",
    "        full_text_xml = fetch_full_text(pmc_id)\n",
    "        plain_text = preprocess_full_text_to_plain_text(full_text_xml)\n",
    "        return plain_text\n",
    "    else:\n",
    "        print(\"Full text not available in PMC for this article.\")\n",
    "        # Return the abstract instead\n",
    "        abstract = df.loc[df['Title'] == df.iloc[0]['Title'], 'Abstract'].values[0]\n",
    "        return abstract"
   ],
   "outputs": [],
   "execution_count": 6
  },
  {
   "cell_type": "code",
   "id": "bcf2f969-2003-4b20-ad7c-a0b93bdecb6e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-30T04:15:17.188029Z",
     "start_time": "2024-08-30T04:15:17.185600Z"
    }
   },
   "source": [
    "def fetch_articles_from_query(query):\n",
    "    ids = pubmed_search(query, retmax=10)  # You can adjust retmax to get more results\n",
    "    \n",
    "    if ids:\n",
    "        details = fetch_details(ids)\n",
    "        df = preprocess_details_to_dataframe(details)\n",
    "\n",
    "        return df, ids\n",
    "    else:\n",
    "        print(\"No results found\")\n",
    "        return None, None"
   ],
   "outputs": [],
   "execution_count": 7
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5157cd5e-ec40-41bf-9c7c-1cadd0406484",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"de Kok et al. 1996;\n",
    "# Ahn et al. 2009; Naranjo et al. 2010\"\n",
    "query = \"deafness Kok AND 1996[dp]\"\n",
    "df, ids = fetch_articles_from_query(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a7ed2b30-7572-4eeb-bcb1-763c62671891",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)  # Show all rows\n",
    "pd.set_option('display.max_columns', None)  # Show all columns\n",
    "pd.set_option('display.max_colwidth', None)  # Show full content of each column\n",
    "pd.set_option('display.width', None)  # Set the display width to be unlimited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7e8653fd-7a7e-4e8b-af9c-9f244ce35bc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Publication Date</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Identification of a hot spot for microdeletions in patients with X-linked deafness type 3 (DFN3) 900 kb proximal to the DFN3 gene POU3F4.</td>\n",
       "      <td>Y J de Kok, E R Vossenaar, C W Cremers, N Dahl, J Laporte, L J Hu, D Lacombe, N Fischel-Ghodsian, R A Friedman, L S Parnes, P Thorpe, M Bitner-Glindzicz, H J Pander, H Heilbronner, J Graveline, J T den Dunnen, H G Brunner, H H Ropers, F P Cremers</td>\n",
       "      <td>Human molecular genetics</td>\n",
       "      <td>1996-Sep-</td>\n",
       "      <td>Small mutations in the POU domain gene POU3F4 were recently shown to cause X-linked deafness type 3 (DFN3) in nine unrelated males. The POU3F4 gene was found to be located outside four of five deletions associated with DFN3. Two of these deletions were situated more than 400 kb proximal to POU3F4. Employing PCR analysis of sequence tagged sites from this region we initially identified novel deletions in two DFN3 patients. To investigate this chromosomal segment in more detail, we extended a previously established 850 kb cosmid contig in the centromeric direction to a total size of 1500 kb. Cosmids from this contig were hybridized to DNA of 11 unrelated males with DFN3. In two patients, we identified deletions encompassing the POU3F4 gene and variably sized segments of Xq21.1. In six of the nine remaining patients which lacked mutations in the POU3F4 gene, smaller deletions were identified which, with one exception, overlap in a 8 kb segment 900 kb proximal to the POU3F4 gene. In one patient, we identified several small deletions in the vicinity of the 8 kb DNA segment. Together, deletions account for 56% (13/23) of all known DFN3 mutations, most (10/13) of which do not encompass the POU3F4 gene. The combined molecular data suggest that the deletion hot spot region in Xq21.1 contains another DFN3 gene or, alternatively, a sequence element involved in transcriptional regulation of POU3F4.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                       Title  \\\n",
       "0  Identification of a hot spot for microdeletions in patients with X-linked deafness type 3 (DFN3) 900 kb proximal to the DFN3 gene POU3F4.   \n",
       "\n",
       "                                                                                                                                                                                                                                                  Authors  \\\n",
       "0  Y J de Kok, E R Vossenaar, C W Cremers, N Dahl, J Laporte, L J Hu, D Lacombe, N Fischel-Ghodsian, R A Friedman, L S Parnes, P Thorpe, M Bitner-Glindzicz, H J Pander, H Heilbronner, J Graveline, J T den Dunnen, H G Brunner, H H Ropers, F P Cremers   \n",
       "\n",
       "                    Journal Publication Date  \\\n",
       "0  Human molecular genetics        1996-Sep-   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            Abstract  \n",
       "0  Small mutations in the POU domain gene POU3F4 were recently shown to cause X-linked deafness type 3 (DFN3) in nine unrelated males. The POU3F4 gene was found to be located outside four of five deletions associated with DFN3. Two of these deletions were situated more than 400 kb proximal to POU3F4. Employing PCR analysis of sequence tagged sites from this region we initially identified novel deletions in two DFN3 patients. To investigate this chromosomal segment in more detail, we extended a previously established 850 kb cosmid contig in the centromeric direction to a total size of 1500 kb. Cosmids from this contig were hybridized to DNA of 11 unrelated males with DFN3. In two patients, we identified deletions encompassing the POU3F4 gene and variably sized segments of Xq21.1. In six of the nine remaining patients which lacked mutations in the POU3F4 gene, smaller deletions were identified which, with one exception, overlap in a 8 kb segment 900 kb proximal to the POU3F4 gene. In one patient, we identified several small deletions in the vicinity of the 8 kb DNA segment. Together, deletions account for 56% (13/23) of all known DFN3 mutations, most (10/13) of which do not encompass the POU3F4 gene. The combined molecular data suggest that the deletion hot spot region in Xq21.1 contains another DFN3 gene or, alternatively, a sequence element involved in transcriptional regulation of POU3F4.  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4eb626bd-18b7-4d59-9c4f-df2bb10eac31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n\\n\\n\\nIntroduction\\n\\nThe genetic basis for multiple Mendelian conditions was initially identified by st'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text = fetch_text_from_pubmed_id(ids[0])\n",
    "text[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a5a3ca07-154b-4cce-b588-6c6ab9805718",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Publication Date</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Characterization of new otic enhancers of the pou3f4 gene reveal distinct signaling pathway regulation and spatio-temporal patterns.</td>\n",
       "      <td>Àlex Robert-Moreno, Silvia Naranjo, Elisa de la Calle-Mustienes, José Luis Gómez-Skarmeta, Berta Alsina</td>\n",
       "      <td>PloS one</td>\n",
       "      <td>2010-Dec-31</td>\n",
       "      <td>POU3F4 is a member of the POU-homedomain transcription factor family with a prominent role in inner ear development. Mutations in the human POU3F4 coding unit leads to X-linked deafness type 3 (DFN3), characterized by conductive hearing loss and progressive sensorineural deafness. Microdeletions found 1 Mb 5' upstream of the coding region also displayed the same phenotype, suggesting that cis-regulatory elements might be present in that region. Indeed, we and others have recently identified several enhancers at the 1 Mb 5' upstream interval of the pou3f4 locus. Here we characterize the spatio-temporal patterns of these regulatory elements in zebrafish transgenic lines. We show that the most distal enhancer (HCNR 81675) is activated earlier and drives GFP reporter expression initially to a broad ear domain to progressively restrict to the sensory patches. The proximal enhancer (HCNR 82478) is switched later during development and promotes expression, among in other tissues, in sensory patches from its onset. The third enhancer (HCNR 81728) is also active at later stages in the otic mesenchyme and in the otic epithelium. We also characterize the signaling pathways regulating these enhancers. While HCNR 81675 is regulated by very early signals of retinoic acid, HCNR 82478 is regulated by Fgf activity at a later stage and the HCNR 81728 enhancer is under the control of Hh signaling. Finally, we show that Sox2 and Pax2 transcription factors are bound to HCNR 81675 genomic region during otic development and specific mutations to these transcription factor binding sites abrogates HCNR 81675 enhancer activity. Altogether, our results suggest that pou3f4 expression in inner ear might be under the control of distinct regulatory elements that fine-tune the spatio-temporal activity of this gene and provides novel data on the signaling mechanisms controlling pou3f4 function.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Multiple enhancers located in a 1-Mb region upstream of POU3F4 promote expression during inner ear development and may be required for hearing.</td>\n",
       "      <td>Silvia Naranjo, Krysta Voesenek, Elisa de la Calle-Mustienes, Alex Robert-Moreno, Haris Kokotas, Maria Grigoriadou, John Economides, Guy Van Camp, Nele Hilgert, Felipe Moreno, Berta Alsina, Michael B Petersen, Hannie Kremer, José Luis Gómez-Skarmeta</td>\n",
       "      <td>Human genetics</td>\n",
       "      <td>2010-Oct-</td>\n",
       "      <td>POU3F4 encodes a POU-domain transcription factor required for inner ear development. Defects in POU3F4 function are associated with X-linked deafness type 3 (DFN3). Multiple deletions affecting up to ~900-kb upstream of POU3F4 are found in DFN3 patients, suggesting the presence of essential POU3F4 enhancers in this region. Recently, an inner ear enhancer was reported that is absent in most DFN3 patients with upstream deletions. However, two indications suggest that additional enhancers in the POU3F4 upstream region are required for POU3F4 function during inner ear development. First, there is at least one DFN3 deletion that does not eliminate the reported enhancer. Second, the expression pattern driven by this enhancer does not fully recapitulate Pou3f4 expression in the inner ear. Here, we screened a 1-Mb region upstream of the POU3F4 gene for additional cis-regulatory elements and searched for novel DFN3 mutations in the identified POU3F4 enhancers. We found several novel enhancers for otic vesicle expression. Some of these also drive expression in kidney, pancreas and brain, tissues that are known to express Pou3f4. In addition, we report a new and smallest deletion identified so far in a DFN3 family which eliminates 3.9 kb, comprising almost exclusively the previous reported inner ear enhancer. We suggest that multiple enhancers control the expression of Pou3f4 in the inner ear and these may contribute to the phenotype observed in DFN3 patients. In addition, the novel deletion demonstrates that the previous reported enhancer, although not sufficient, is essential for POU3F4 function during inner ear development.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                             Title  \\\n",
       "0             Characterization of new otic enhancers of the pou3f4 gene reveal distinct signaling pathway regulation and spatio-temporal patterns.   \n",
       "1  Multiple enhancers located in a 1-Mb region upstream of POU3F4 promote expression during inner ear development and may be required for hearing.   \n",
       "\n",
       "                                                                                                                                                                                                                                                     Authors  \\\n",
       "0                                                                                                                                                    Àlex Robert-Moreno, Silvia Naranjo, Elisa de la Calle-Mustienes, José Luis Gómez-Skarmeta, Berta Alsina   \n",
       "1  Silvia Naranjo, Krysta Voesenek, Elisa de la Calle-Mustienes, Alex Robert-Moreno, Haris Kokotas, Maria Grigoriadou, John Economides, Guy Van Camp, Nele Hilgert, Felipe Moreno, Berta Alsina, Michael B Petersen, Hannie Kremer, José Luis Gómez-Skarmeta   \n",
       "\n",
       "          Journal Publication Date  \\\n",
       "0        PloS one      2010-Dec-31   \n",
       "1  Human genetics        2010-Oct-   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Abstract  \n",
       "0  POU3F4 is a member of the POU-homedomain transcription factor family with a prominent role in inner ear development. Mutations in the human POU3F4 coding unit leads to X-linked deafness type 3 (DFN3), characterized by conductive hearing loss and progressive sensorineural deafness. Microdeletions found 1 Mb 5' upstream of the coding region also displayed the same phenotype, suggesting that cis-regulatory elements might be present in that region. Indeed, we and others have recently identified several enhancers at the 1 Mb 5' upstream interval of the pou3f4 locus. Here we characterize the spatio-temporal patterns of these regulatory elements in zebrafish transgenic lines. We show that the most distal enhancer (HCNR 81675) is activated earlier and drives GFP reporter expression initially to a broad ear domain to progressively restrict to the sensory patches. The proximal enhancer (HCNR 82478) is switched later during development and promotes expression, among in other tissues, in sensory patches from its onset. The third enhancer (HCNR 81728) is also active at later stages in the otic mesenchyme and in the otic epithelium. We also characterize the signaling pathways regulating these enhancers. While HCNR 81675 is regulated by very early signals of retinoic acid, HCNR 82478 is regulated by Fgf activity at a later stage and the HCNR 81728 enhancer is under the control of Hh signaling. Finally, we show that Sox2 and Pax2 transcription factors are bound to HCNR 81675 genomic region during otic development and specific mutations to these transcription factor binding sites abrogates HCNR 81675 enhancer activity. Altogether, our results suggest that pou3f4 expression in inner ear might be under the control of distinct regulatory elements that fine-tune the spatio-temporal activity of this gene and provides novel data on the signaling mechanisms controlling pou3f4 function.  \n",
       "1                                                                                                                                                                                                                                                             POU3F4 encodes a POU-domain transcription factor required for inner ear development. Defects in POU3F4 function are associated with X-linked deafness type 3 (DFN3). Multiple deletions affecting up to ~900-kb upstream of POU3F4 are found in DFN3 patients, suggesting the presence of essential POU3F4 enhancers in this region. Recently, an inner ear enhancer was reported that is absent in most DFN3 patients with upstream deletions. However, two indications suggest that additional enhancers in the POU3F4 upstream region are required for POU3F4 function during inner ear development. First, there is at least one DFN3 deletion that does not eliminate the reported enhancer. Second, the expression pattern driven by this enhancer does not fully recapitulate Pou3f4 expression in the inner ear. Here, we screened a 1-Mb region upstream of the POU3F4 gene for additional cis-regulatory elements and searched for novel DFN3 mutations in the identified POU3F4 enhancers. We found several novel enhancers for otic vesicle expression. Some of these also drive expression in kidney, pancreas and brain, tissues that are known to express Pou3f4. In addition, we report a new and smallest deletion identified so far in a DFN3 family which eliminates 3.9 kb, comprising almost exclusively the previous reported inner ear enhancer. We suggest that multiple enhancers control the expression of Pou3f4 in the inner ear and these may contribute to the phenotype observed in DFN3 patients. In addition, the novel deletion demonstrates that the previous reported enhancer, although not sufficient, is essential for POU3F4 function during inner ear development.  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# \"de Kok et al. 1996;\n",
    "# Ahn et al. 2009; Naranjo et al. 2010\"\n",
    "query = \"pou3f4 Naranjo AND 2010[dp]\"\n",
    "prompt_df, ids = fetch_articles_from_query(query)\n",
    "prompt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "489c53b7-0d1d-41ee-8a44-d9cbf1caa0f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Publication Date</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Otic mesenchyme expression of Cre recombinase directed by the inner ear enhancer of the Brn4/Pou3f4 gene.</td>\n",
       "      <td>Kyung J Ahn, Frank Passero, E Bryan Crenshaw</td>\n",
       "      <td>Genesis (New York, N.Y. : 2000)</td>\n",
       "      <td>2009-Mar-</td>\n",
       "      <td>Brn4/Pou3f4 is a POU-domain transcription factor expressed in the otic mesenchyme that is required for the normal development of the inner ear. In this report, we describe the isolation of an otic mesenchyme enhancer in the Brn4 gene. Subsequently, this enhancer was used to drive the expression of Cre recombinase in the otic mesenchyme of transgenic mice. When intercrossed with the ROSA reporter strain, R26R, ss-galactosidase expression is detected in several inner ear structures derived from otic mesenchyme, including the temporal bone, spiral ligament, spiral limbus, and mesenchyme underlying sensory epithelium of the utricle, saccule and semicircular canals. Thus, this Cre pedigree can induce conditional rearrangement of genes in the otic mesenchyme, and will serve as a powerful genetic tool to characterize the function of genes in the mesenchymal tissues of the inner ear.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                       Title  \\\n",
       "0  Otic mesenchyme expression of Cre recombinase directed by the inner ear enhancer of the Brn4/Pou3f4 gene.   \n",
       "\n",
       "                                        Authors  \\\n",
       "0  Kyung J Ahn, Frank Passero, E Bryan Crenshaw   \n",
       "\n",
       "                           Journal Publication Date  \\\n",
       "0  Genesis (New York, N.Y. : 2000)        2009-Mar-   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Abstract  \n",
       "0  Brn4/Pou3f4 is a POU-domain transcription factor expressed in the otic mesenchyme that is required for the normal development of the inner ear. In this report, we describe the isolation of an otic mesenchyme enhancer in the Brn4 gene. Subsequently, this enhancer was used to drive the expression of Cre recombinase in the otic mesenchyme of transgenic mice. When intercrossed with the ROSA reporter strain, R26R, ss-galactosidase expression is detected in several inner ear structures derived from otic mesenchyme, including the temporal bone, spiral ligament, spiral limbus, and mesenchyme underlying sensory epithelium of the utricle, saccule and semicircular canals. Thus, this Cre pedigree can induce conditional rearrangement of genes in the otic mesenchyme, and will serve as a powerful genetic tool to characterize the function of genes in the mesenchymal tissues of the inner ear.  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"pou3f4 Ahn AND 2009[dp]\"\n",
    "prompt_df, ids = fetch_articles_from_query(query)\n",
    "prompt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a749f6b2-4d09-49ec-bbff-af007587108b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# Load the CSV file into a pandas DataFrame\n",
    "file_path = '~/Documents/Badawcze/enhancerlit/enhancer_candidates.csv'\n",
    "df = pd.read_csv(file_path)\n",
    "\n",
    "# Function to create queries from the Literature column\n",
    "def create_queries(row):\n",
    "    gene_name = re.sub(r'[^a-zA-Z0-9]', '', row['Gene name'].lower())  # Remove non-alphanumeric characters\n",
    "    queries = []\n",
    "    if pd.notna(row['Literature']):\n",
    "        for ref in row['Literature'].split(';'):\n",
    "            ref = ref.strip()\n",
    "            parts = ref.split()\n",
    "            if len(parts) >= 3:\n",
    "                # Remove 'et al.' and 'i wsp.' from the author's name\n",
    "                if 'et' in parts and 'al.' in parts:\n",
    "                    etal_index = parts.index('et')\n",
    "                    first_author_surname = ' '.join(parts[:etal_index])\n",
    "                elif 'i' in parts and 'wsp.' in parts:\n",
    "                    iwsp_index = parts.index('i')\n",
    "                    first_author_surname = ' '.join(parts[:iwsp_index])\n",
    "                else:\n",
    "                    first_author_surname = ' '.join(parts[:-1])  # Include all parts except the last one as the surname\n",
    "                \n",
    "                year_of_publication = parts[-1].strip(';,')\n",
    "                query = f\"{gene_name} {first_author_surname} AND {year_of_publication}[dp]\"\n",
    "                queries.append(query)\n",
    "    return queries\n",
    "\n",
    "# Apply the function to each row in the DataFrame\n",
    "df['Queries'] = df.apply(create_queries, axis=1)\n",
    "\n",
    "# Extract the list of lists of queries\n",
    "list_of_queries = df['Queries'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "767c7dc3-7350-4220-ab87-be0e830dce19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['pou3f4 de Kok AND 1996[dp]',\n",
       "  'pou3f4 Ahn AND 2009[dp]',\n",
       "  'pou3f4 Naranjo AND 2010[dp]'],\n",
       " ['sost Balemans AND 2002[dp]', 'sost Loots AND 2005[dp]'],\n",
       " ['shox Sabherwal AND 2007[dp]',\n",
       "  'shox Fukami AND 2005[dp]',\n",
       "  'shox Benito-Sanz AND 2012[dp]'],\n",
       " ['nr0b1 Smyk AND 2007b[dp]'],\n",
       " ['sry McElreavy AND 1992[dp]',\n",
       "  'sry Capel AND 1993[dp]',\n",
       "  'sry McElreavy AND 1996[dp]',\n",
       "  'sry Sharp AND 2005[dp]'],\n",
       " ['hbb Driscoll AND 1989[dp]'],\n",
       " ['hba1hba2 Viprakasit AND 2006[dp]'],\n",
       " ['pitx2 Trembath AND 2004[dp]'],\n",
       " ['foxc1 Davies AND 1999[dp]'],\n",
       " ['foxl2 Beysen AND 2005[dp]', \"foxl2 D'haene AND 2009[dp]\"],\n",
       " ['sox9 Benko AND 2009[dp]',\n",
       "  'sox9 Fukami AND 2012[dp]',\n",
       "  'sox9 Sanchez-Castro AND 2013[dp]'],\n",
       " ['sox9 Kim AND 2015[dp]'],\n",
       " ['pitx2 Volkmann AND 2011[dp]'],\n",
       " ['atoh7 Ghiasvand AND 2011[dp]'],\n",
       " ['dlx6dlx5 Brown AND 2010[dp]'],\n",
       " ['dlx6dlx5 Tayebi AND 2014[dp]'],\n",
       " ['pax6 Wawrocka AND 2012[dp]'],\n",
       " ['foxg1 Allou AND 2012[dp]'],\n",
       " ['foxf1 Szafranski AND 2013[dp]'],\n",
       " ['nr0b1 Skinningsrud AND 2009[dp]'],\n",
       " ['hbb Kioussis AND 1983[dp]'],\n",
       " ['pax6 Fantes AND 1995[dp]', 'pax6 Kleinjan AND 2001[dp]'],\n",
       " ['twist Cai AND 2003[dp]'],\n",
       " ['pitx2 Flomen AND 1997[dp]'],\n",
       " ['pitx1 Spielmann AND 2012[dp]'],\n",
       " ['gli3 Vortkamp AND 1991[dp]'],\n",
       " ['maf Jamieson AND 2002[dp]'],\n",
       " ['foxc2 Fang AND 2000[dp]'],\n",
       " ['foxl2 Crisponi AND 2004[dp]'],\n",
       " ['foxp2 Scherer AND 2003[dp]'],\n",
       " ['tgfb2 David AND 2003[dp]'],\n",
       " ['sox9 Leipoldt AND 2007[dp]', 'sox9 Velagaleti AND 2005[dp]'],\n",
       " ['sox9 Benko AND 2009[dp]', 'sox9 Jakobsen AND 2007[dp]'],\n",
       " ['plp1 Muncke AND 2004[dp]'],\n",
       " ['alx4ext2 Wakui AND 2005[dp]'],\n",
       " ['satb2 FitzPatrick AND 2003[dp]'],\n",
       " ['sall1 Marlin AND 1999[dp]'],\n",
       " ['six3 Wallis AND 1999[dp]'],\n",
       " ['runx2 Fernandez AND 2005[dp]'],\n",
       " ['reep3 Castermans AND 2007[dp]'],\n",
       " ['sox3 Bowl AND 2005[dp]'],\n",
       " ['shh Roessler AND 1997[dp]', 'shh Fernandez AND 2005[dp]'],\n",
       " ['hoxd131 Spitz AND 2002[dp]', 'hoxd131 Dlugaszewska AND 2006[dp]'],\n",
       " ['dlx6dlx5 Brown AND 2010[dp]'],\n",
       " ['shh Lettice AND 2002[dp]'],\n",
       " ['lct Enattah AND 2002[dp]'],\n",
       " ['irf6 Rahimov AND 2008[dp]'],\n",
       " ['sox9 Benko AND 2009[dp]'],\n",
       " ['shh Jeong AND 2008[dp]'],\n",
       " ['shh Lettice AND 2012[dp]'],\n",
       " ['pax6 Bhatia AND 2013[dp]'],\n",
       " ['ptf1a Weedon AND 2014[dp]'],\n",
       " ['dlx6dlx5 Poitras AND 2010[dp]'],\n",
       " ['fgfr2 Leslie AND 2015[dp]'],\n",
       " ['hba1hba2 Tufarelli AND 2003[dp]'],\n",
       " ['msh2 Ligtenberg AND 2009[dp]'],\n",
       " ['hba1hba2 de Gobbi AND 2006[dp]'],\n",
       " ['dux4 Wijmenga AND 1992[dp]',\n",
       "  'dux4 van Deutekom AND 1993[dp]',\n",
       "  'dux4 Ottaviani AND 2009[dp]',\n",
       "  'dux4 Snider AND 2009[dp]'],\n",
       " ['shh Klopocki AND 2008[dp]', 'shh Sun M AND 2008[dp]'],\n",
       " ['sox9 Kim AND 2015[dp]', 'sox9 Hyon AND 2015[dp]'],\n",
       " ['plp1 Lee AND 2006[dp]'],\n",
       " ['ihh Klopocki AND 2011[dp]'],\n",
       " ['sox9 Kurth AND 2009[dp]'],\n",
       " ['bmp2 Dathe AND 2009[dp]', 'bmp2 Su AND 2011[dp]'],\n",
       " ['shh Klopocki AND 2008[dp]'],\n",
       " ['cyp19 Demura AND 2007[dp]'],\n",
       " ['myc Erikson AND 1983[dp]', 'myc Hayday AND 1984[dp]']]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the resulting list of lists of queries\n",
    "list_of_queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4d0d9dbd-5736-4d37-94b8-ce8990c4483d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pou3f4 de Kok AND 1996[dp]',\n",
       " 'pou3f4 Ahn AND 2009[dp]',\n",
       " 'pou3f4 Naranjo AND 2010[dp]']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_queries[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "bd6a1dcb-fb39-44c4-abd5-94b609aece7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "research_id = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6c34993f-6607-49c7-8f8a-1ca24ccf9442",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['8872461', '19217071', '21209840', '20668882']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids_for_row_2d = [fetch_articles_from_query(q)[research_id] for q in list_of_queries[0]]\n",
    "# Flatten\n",
    "ids_for_row = [query for sublist in ids_for_row_2d for query in sublist]\n",
    "\n",
    "ids_for_row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "49383a5d-becf-4304-9497-e6e9d1b1bce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "texts = [fetch_text_from_pubmed_id(id) for id in ids_for_row]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83b14efa-b547-4c3f-8342-722367d05064",
   "metadata": {},
   "source": [
    "## Retreive metadata of searches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1aa04f01-57f5-46ae-a0f0-f906e3ede213",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_string_list(row):\n",
    "    string_list = []\n",
    "    for column in row.index[:-3]:  # Exclude the last two columns\n",
    "        value = row[column]\n",
    "        # Ensure that NaN values are handled appropriately\n",
    "        value_str = str(value) if pd.notna(value) else \"brak danych\"\n",
    "        string_list.append(f\"{column}: {value_str}\")\n",
    "    return string_list\n",
    "\n",
    "searches_metadata = df.apply(generate_string_list, axis=1).tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3eb8987c-f96a-4934-90b0-519b60789fb5",
   "metadata": {},
   "source": [
    "# Answering the question about ClinVar and coordinates\n",
    "Having the full texts of the biomedical articles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af3fdd70-12ee-449c-88cf-dca0edef39b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install openai==0.28"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6edcfb4f-0b44-41fb-94d1-666b1935c47e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import openai\n",
    "from langchain.prompts import PromptTemplate\n",
    "\n",
    "# Set your OpenAI API key\n",
    "openai.api_key = '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'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "19c863ed-a381-4fc0-b7e7-6f3e6a5d563a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.28.0'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "openai.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1b9b299a-ad80-4aa2-8ff5-3c72b5eceac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the context strings\n",
    "context_strings = [f\"CONTEXT INFO: Publication {i}: {text}\" for i, text in enumerate(texts)]\n",
    "context_strings.extend([f\"CONTEXT INFO: {x}\" for x in searches_metadata[research_id]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b95a0db-01df-4ebf-9a65-abff6c1286ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4c33757-2a70-4435-9b04-ef82d2958a51",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7507578f-702c-4b35-b7d5-2c60890639d4",
   "metadata": {},
   "outputs": [
    {
     "ename": "InvalidRequestError",
     "evalue": "This model's maximum context length is 16385 tokens. However, your messages resulted in 50218 tokens. Please reduce the length of the messages.",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mInvalidRequestError\u001B[0m                       Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[44], line 28\u001B[0m\n\u001B[1;32m     25\u001B[0m final_prompt \u001B[38;5;241m=\u001B[39m create_prompt(context, question)\n\u001B[1;32m     27\u001B[0m \u001B[38;5;66;03m# Call the OpenAI API with the final prompt using the new interface\u001B[39;00m\n\u001B[0;32m---> 28\u001B[0m response \u001B[38;5;241m=\u001B[39m \u001B[43mopenai\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mChatCompletion\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mcreate\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m     29\u001B[0m \u001B[43m    \u001B[49m\u001B[43mmodel\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mgpt-3.5-turbo\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m     30\u001B[0m \u001B[43m    \u001B[49m\u001B[43mmessages\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43m[\u001B[49m\n\u001B[1;32m     31\u001B[0m \u001B[43m        \u001B[49m\u001B[43m{\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mrole\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43msystem\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mcontent\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mYou are a helpful assistant.\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m}\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m     32\u001B[0m \u001B[43m        \u001B[49m\u001B[43m{\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mrole\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43muser\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mcontent\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[43mfinal_prompt\u001B[49m\u001B[43m}\u001B[49m\n\u001B[1;32m     33\u001B[0m \u001B[43m    \u001B[49m\u001B[43m]\u001B[49m\n\u001B[1;32m     34\u001B[0m \u001B[43m)\u001B[49m\n\u001B[1;32m     36\u001B[0m \u001B[38;5;66;03m# Print the response\u001B[39;00m\n\u001B[1;32m     37\u001B[0m \u001B[38;5;28mprint\u001B[39m(response\u001B[38;5;241m.\u001B[39mchoices[\u001B[38;5;241m0\u001B[39m]\u001B[38;5;241m.\u001B[39mmessage[\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mcontent\u001B[39m\u001B[38;5;124m\"\u001B[39m])\n",
      "File \u001B[0;32m~/anaconda3/envs/enhancerlit/lib/python3.11/site-packages/openai/api_resources/chat_completion.py:25\u001B[0m, in \u001B[0;36mChatCompletion.create\u001B[0;34m(cls, *args, **kwargs)\u001B[0m\n\u001B[1;32m     23\u001B[0m \u001B[38;5;28;01mwhile\u001B[39;00m \u001B[38;5;28;01mTrue\u001B[39;00m:\n\u001B[1;32m     24\u001B[0m     \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m---> 25\u001B[0m         \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43msuper\u001B[39;49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mcreate\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m     26\u001B[0m     \u001B[38;5;28;01mexcept\u001B[39;00m TryAgain \u001B[38;5;28;01mas\u001B[39;00m e:\n\u001B[1;32m     27\u001B[0m         \u001B[38;5;28;01mif\u001B[39;00m timeout \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m \u001B[38;5;129;01mand\u001B[39;00m time\u001B[38;5;241m.\u001B[39mtime() \u001B[38;5;241m>\u001B[39m start \u001B[38;5;241m+\u001B[39m timeout:\n",
      "File \u001B[0;32m~/anaconda3/envs/enhancerlit/lib/python3.11/site-packages/openai/api_resources/abstract/engine_api_resource.py:153\u001B[0m, in \u001B[0;36mEngineAPIResource.create\u001B[0;34m(cls, api_key, api_base, api_type, request_id, api_version, organization, **params)\u001B[0m\n\u001B[1;32m    127\u001B[0m \u001B[38;5;129m@classmethod\u001B[39m\n\u001B[1;32m    128\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mcreate\u001B[39m(\n\u001B[1;32m    129\u001B[0m     \u001B[38;5;28mcls\u001B[39m,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m    136\u001B[0m     \u001B[38;5;241m*\u001B[39m\u001B[38;5;241m*\u001B[39mparams,\n\u001B[1;32m    137\u001B[0m ):\n\u001B[1;32m    138\u001B[0m     (\n\u001B[1;32m    139\u001B[0m         deployment_id,\n\u001B[1;32m    140\u001B[0m         engine,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m    150\u001B[0m         api_key, api_base, api_type, api_version, organization, \u001B[38;5;241m*\u001B[39m\u001B[38;5;241m*\u001B[39mparams\n\u001B[1;32m    151\u001B[0m     )\n\u001B[0;32m--> 153\u001B[0m     response, _, api_key \u001B[38;5;241m=\u001B[39m \u001B[43mrequestor\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mrequest\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m    154\u001B[0m \u001B[43m        \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mpost\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m    155\u001B[0m \u001B[43m        \u001B[49m\u001B[43murl\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    156\u001B[0m \u001B[43m        \u001B[49m\u001B[43mparams\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mparams\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    157\u001B[0m \u001B[43m        \u001B[49m\u001B[43mheaders\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mheaders\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    158\u001B[0m \u001B[43m        \u001B[49m\u001B[43mstream\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstream\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    159\u001B[0m \u001B[43m        \u001B[49m\u001B[43mrequest_id\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mrequest_id\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    160\u001B[0m \u001B[43m        \u001B[49m\u001B[43mrequest_timeout\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mrequest_timeout\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    161\u001B[0m \u001B[43m    \u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    163\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m stream:\n\u001B[1;32m    164\u001B[0m         \u001B[38;5;66;03m# must be an iterator\u001B[39;00m\n\u001B[1;32m    165\u001B[0m         \u001B[38;5;28;01massert\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(response, OpenAIResponse)\n",
      "File \u001B[0;32m~/anaconda3/envs/enhancerlit/lib/python3.11/site-packages/openai/api_requestor.py:298\u001B[0m, in \u001B[0;36mAPIRequestor.request\u001B[0;34m(self, method, url, params, headers, files, stream, request_id, request_timeout)\u001B[0m\n\u001B[1;32m    277\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mrequest\u001B[39m(\n\u001B[1;32m    278\u001B[0m     \u001B[38;5;28mself\u001B[39m,\n\u001B[1;32m    279\u001B[0m     method,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m    286\u001B[0m     request_timeout: Optional[Union[\u001B[38;5;28mfloat\u001B[39m, Tuple[\u001B[38;5;28mfloat\u001B[39m, \u001B[38;5;28mfloat\u001B[39m]]] \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[1;32m    287\u001B[0m ) \u001B[38;5;241m-\u001B[39m\u001B[38;5;241m>\u001B[39m Tuple[Union[OpenAIResponse, Iterator[OpenAIResponse]], \u001B[38;5;28mbool\u001B[39m, \u001B[38;5;28mstr\u001B[39m]:\n\u001B[1;32m    288\u001B[0m     result \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mrequest_raw(\n\u001B[1;32m    289\u001B[0m         method\u001B[38;5;241m.\u001B[39mlower(),\n\u001B[1;32m    290\u001B[0m         url,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m    296\u001B[0m         request_timeout\u001B[38;5;241m=\u001B[39mrequest_timeout,\n\u001B[1;32m    297\u001B[0m     )\n\u001B[0;32m--> 298\u001B[0m     resp, got_stream \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_interpret_response\u001B[49m\u001B[43m(\u001B[49m\u001B[43mresult\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mstream\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    299\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m resp, got_stream, \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mapi_key\n",
      "File \u001B[0;32m~/anaconda3/envs/enhancerlit/lib/python3.11/site-packages/openai/api_requestor.py:700\u001B[0m, in \u001B[0;36mAPIRequestor._interpret_response\u001B[0;34m(self, result, stream)\u001B[0m\n\u001B[1;32m    692\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m (\n\u001B[1;32m    693\u001B[0m         \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_interpret_response_line(\n\u001B[1;32m    694\u001B[0m             line, result\u001B[38;5;241m.\u001B[39mstatus_code, result\u001B[38;5;241m.\u001B[39mheaders, stream\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m\n\u001B[1;32m    695\u001B[0m         )\n\u001B[1;32m    696\u001B[0m         \u001B[38;5;28;01mfor\u001B[39;00m line \u001B[38;5;129;01min\u001B[39;00m parse_stream(result\u001B[38;5;241m.\u001B[39miter_lines())\n\u001B[1;32m    697\u001B[0m     ), \u001B[38;5;28;01mTrue\u001B[39;00m\n\u001B[1;32m    698\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m    699\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m (\n\u001B[0;32m--> 700\u001B[0m         \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_interpret_response_line\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m    701\u001B[0m \u001B[43m            \u001B[49m\u001B[43mresult\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mcontent\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mdecode\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mutf-8\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    702\u001B[0m \u001B[43m            \u001B[49m\u001B[43mresult\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mstatus_code\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    703\u001B[0m \u001B[43m            \u001B[49m\u001B[43mresult\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mheaders\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m    704\u001B[0m \u001B[43m            \u001B[49m\u001B[43mstream\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;28;43;01mFalse\u001B[39;49;00m\u001B[43m,\u001B[49m\n\u001B[1;32m    705\u001B[0m \u001B[43m        \u001B[49m\u001B[43m)\u001B[49m,\n\u001B[1;32m    706\u001B[0m         \u001B[38;5;28;01mFalse\u001B[39;00m,\n\u001B[1;32m    707\u001B[0m     )\n",
      "File \u001B[0;32m~/anaconda3/envs/enhancerlit/lib/python3.11/site-packages/openai/api_requestor.py:765\u001B[0m, in \u001B[0;36mAPIRequestor._interpret_response_line\u001B[0;34m(self, rbody, rcode, rheaders, stream)\u001B[0m\n\u001B[1;32m    763\u001B[0m stream_error \u001B[38;5;241m=\u001B[39m stream \u001B[38;5;129;01mand\u001B[39;00m \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124merror\u001B[39m\u001B[38;5;124m\"\u001B[39m \u001B[38;5;129;01min\u001B[39;00m resp\u001B[38;5;241m.\u001B[39mdata\n\u001B[1;32m    764\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m stream_error \u001B[38;5;129;01mor\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;241m200\u001B[39m \u001B[38;5;241m<\u001B[39m\u001B[38;5;241m=\u001B[39m rcode \u001B[38;5;241m<\u001B[39m \u001B[38;5;241m300\u001B[39m:\n\u001B[0;32m--> 765\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mhandle_error_response(\n\u001B[1;32m    766\u001B[0m         rbody, rcode, resp\u001B[38;5;241m.\u001B[39mdata, rheaders, stream_error\u001B[38;5;241m=\u001B[39mstream_error\n\u001B[1;32m    767\u001B[0m     )\n\u001B[1;32m    768\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m resp\n",
      "\u001B[0;31mInvalidRequestError\u001B[0m: This model's maximum context length is 16385 tokens. However, your messages resulted in 50218 tokens. Please reduce the length of the messages."
     ]
    }
   ],
   "source": [
    "# Combine the context strings into a single context\n",
    "context = \"\\n\".join(context_strings)\n",
    "\n",
    "# Define the prompt template with the context\n",
    "prompt_template = PromptTemplate(\n",
    "    input_variables=[\"context\", \"question\"],\n",
    "    template=\"\"\"\n",
    "    Context:\n",
    "    {context}\n",
    "    \n",
    "    Question:\n",
    "    {question}\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "# Create the final prompt using the template\n",
    "def create_prompt(context, question):\n",
    "    return prompt_template.format(context=context, question=question)\n",
    "\n",
    "# Define the question you want to ask\n",
    "# question = \"What are the related ClinVar variant ids in this context?\"\n",
    "question = \"What are the exact coordinates of this enhancer?\"\n",
    "\n",
    "# Create the final prompt\n",
    "final_prompt = create_prompt(context, question)\n",
    "\n",
    "# Call the OpenAI API with the final prompt using the new interface\n",
    "response = openai.ChatCompletion.create(\n",
    "    model=\"gpt-3.5-turbo\",\n",
    "    messages=[\n",
    "        {\"role\": \"system\", \"content\": \"You are a helpful assistant.\"},\n",
    "        {\"role\": \"user\", \"content\": final_prompt}\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Print the response\n",
    "print(response.choices[0].message[\"content\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f656b023-69fe-463d-85f7-45a1dbc52107",
   "metadata": {},
   "source": [
    "### Testing BiomedBERT (formerly: PubMedBERT)\n",
    "Demands fine-tuning on a downstream task"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe6197e5-d847-4153-8854-cd183212f056",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "from transformers import AutoModelForTokenClassification, AutoTokenizer\n",
    "\n",
    "# Check if GPU is available\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "# Load the pre-trained GENE-BERT model and tokenizer from Hugging Face\n",
    "model_name = \"microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext\"  # Replace with the actual model name if available on Hugging Face\n",
    "model = AutoModelForTokenClassification.from_pretrained(model_name).to(device)\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "\n",
    "# Function to perform inference\n",
    "def predict(text):\n",
    "    inputs = tokenizer(text, return_tensors=\"pt\").to(device)\n",
    "    print(inputs)\n",
    "    with torch.no_grad():\n",
    "        outputs = model(**inputs)\n",
    "    \n",
    "    # Assuming the model is for token classification, extract predictions\n",
    "    logits = outputs.logits\n",
    "    predictions = torch.argmax(logits, dim=-1).cpu().numpy()\n",
    "    \n",
    "    # Convert predictions to human-readable labels (assuming a label map is available)\n",
    "    label_map = {i: label for i, label in enumerate(tokenizer.get_vocab().keys())}\n",
    "    predicted_labels = [label_map[prediction] for prediction in predictions[0]]\n",
    "    \n",
    "    return predicted_labels\n",
    "\n",
    "# Example usage\n",
    "text = \"Example genomic text for testing GENE-BERT.\"\n",
    "predictions = predict(text)\n",
    "print(\"Predictions:\", predictions)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "29545882-d50a-45f9-b5d1-3be536ca6bf0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input_ids': tensor([[   2,    4, 1977,   43, 2798, 9241, 2359,   18,    3]],\n",
      "       device='cuda:0'), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0]], device='cuda:0'), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0')}\n",
      "Predictions: ['lesion', 'lesion', 'lesion', 'lesion', 'lesion', 'lesion', 'lesion', 'linkage', 'linkage']\n"
     ]
    }
   ],
   "source": [
    "text = \"\"\n",
    "predictions = predict(text)\n",
    "print(\"Predictions:\", predictions)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b5ac2a1-cfc9-4b61-8bef-6fad7e659fa4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
